packages/gatsby/src/utils/page-data.js

```diff
@@ -35,7 +40,71 @@ const write = async ({ publicDir }, page, result) => {
   await fs.outputFile(filePath, bodyStr)
 }
 
+const getChangedPageDataKeys = (state, cachedPageData) => {
+  if (cachedPageData && state.pageData) {
+    const pageKeys = []
+    state.pageData.forEach((resultHash, key) => {
+      if (!cachedPageData.has(key)) {
+        pageKeys.push(key)
+      } else {
+        const newPageDataHash = resultHash
+        const previousPageDataHash = cachedPageData.get(key)
+
+        if (newPageDataHash !== previousPageDataHash) {
+          pageKeys.push(key)
+        }
+      }
+    })
+    return pageKeys
+  }
+
+  return [...state.pages.keys()]
+}
+
+const collectRemovedPageData = (state, cachedPageData) => {
+  if (cachedPageData && state.pageData) {
+    const deletedPageKeys = []
+    cachedPageData.forEach((_value, key) => {
+      if (!state.pageData.has(key)) {
+        deletedPageKeys.push(key)
+      }
+    })
+    return deletedPageKeys
+  }
+  return []
+}
+
+const checkAndRemoveEmptyDir = (dir, pagePath) => {
+  const hasFiles = fs.readdirSync(path.join(dir, pagePath), `utf8`, true)
+  if (!hasFiles.length) {
+    fs.removeSync(path.join(dir, pagePath))
+  }
+}
+
+const removePageFiles = ({ publicDir }, pageKeys) => {
+  const removePages = pageKeys.map(pagePath => {
+    const pageHtmlFile = getFilePath({ publicDir }, pagePath, `html`)
+    return fs
+      .remove(pageHtmlFile)
+      .then(() => checkAndRemoveEmptyDir(publicDir, pagePath))
```

Comment:
This might still be tricky - it might or might not work depending on the order of `pageKeys` + timing of `fs.remove` execution:

If order will be [`/blog`, `/blog/something`] (and we didn't have any other `/blog/*` subpages)

We would call `fs.remove` with `public/blog/index.html` and then `public/blog/something/index.html`

If timing of those functions would be so we get to `.then` for `/blog` first - then the `public/blog` would still be no empty (because `public/blog/something` still exist at that point), so we wouldn't remove `public/blog` because of timing/ordering.

That's why I think we potentially would need something like this maybe?

Make 2 separate passes over `pageKeys` (for html files):
1. iterate over `pageKeys` and remove just html files and await on all of it to finish
2. iterate over `pageKeys` and remove directories - now this is still not great, because our order still will impact result, so perhaps we would need to sort `pageKeys` first so the most nested paths are handled first ? (so we should sort pageKeys by amount of slashes that they contain to ensure we always handle `public/blog/something` before `public/blog`


I think it's the same for `page-data` loop below, but in case of `page-data` we shouldn't be affected by existence of empty directories in the runtime ( as opposed to empty directories that usually contain .html files where hosting might get confused if there is empty dir )

Reply:
@pieh  🙇  Yep that makes sense 
So after the `Promise.all` I've done `forEach` on a sorted list of page keys. This should remove from most nested directories to least.
This should sort out the ordering issue 🤞